<!-- This launch file launches all nodes necessary for autonomous navigation. -->
<launch>
    <arg name="run_tag_detector" default="true"/>
    <arg name="sim" default="false"/>
    <arg name="use_ekf" default="true"/>
    <arg name="ekf_start_delay" default="0"/>
    <arg name="use_odom_frame" default="true"/>

    

    <!--
      ==========
      Perception
      ==========
    -->
    <node pkg="nodelet" type="nodelet" name="perception_nodelet_manager" respawn="true"
          args="manager" output="screen"/>
    <!-- nodelet to detect AR tags and publish them to the TF tree -->
    <node if="$(arg run_tag_detector)"
          pkg="nodelet" type="nodelet" name="tag_detector" respawn="true"
          args="load mrover/TagDetectorNodelet perception_nodelet_manager" output="screen"/>

    <!--
      ===========
      Navigation
      ===========
    -->
    <!-- node to navigate to a series of waypoints, AR tags, and gates -->
    <rosparam command="load" file="$(find mrover)/config/navigation.yaml"/>
    <node name="nav" pkg="mrover" type="navigation.py"/>
    <node name="failure_id" pkg="mrover" type="failure_identification.py"/>

    <!--
      ============
      Localization
      ============
    -->
    <rosparam command="load" file="$(find mrover)/config/localization.yaml"/>
    <rosparam if="$(arg sim)" command="load" file="$(find mrover)/config/sim_ekf.yaml"/>
    <rosparam unless="$(arg sim)" command="load" file="$(find mrover)/config/ekf.yaml"/>

    <!-- global EKF -->
    <node if="$(arg use_ekf)" pkg="robot_localization" type="ekf_localization_node" name="global_ekf"
          clear_params="true" launch-prefix="bash -c 'sleep $(arg ekf_start_delay); $0 $@'">
        <remap from="odometry/filtered" to="global_ekf/odometry"/>
    </node>

    <!-- passthrough filter to replace EKF in datapath -->
    <node unless="$(arg use_ekf)" pkg="mrover" type="passthrough_filter.py" name="passthrough_filter"/>

    <!-- node to linearize GPS from geodetic to ENU cartesian coords -->
    <node name="gps_linearization" pkg="mrover" type="gps_linearization.py" output="screen"/>

    <!-- plot the rover trajectory in rviz as it drives -->
    <!-- <node name="hector_trajectory_server" output="screen" pkg="hector_trajectory_server" type="hector_trajectory_server">
        <param name="target_frame_name" type="string" value="map"/>
        <param name="source_frame_name" type="string" value="base_link"/>
        <param name="trajectory_update_rate" type="double" value="4"/>
        <param name="trajectory_publish_rate" type="double" value="4"/>
    </node> -->
    <!--
      Set rosparams required for odom frame if its true, else set them correclty to not use odom frame
    -->
    <param if="$(arg use_odom_frame)" name="use_odom_frame" value="true"/>
    <param unless="$(arg use_odom_frame)" name="use_odom_frame" value="false"/>
    <param if="$(arg use_odom_frame)" name="odom_frame" value="odom"/>
    <param unless="$(arg use_odom_frame)" name="odom_frame" value="map"/>
    <param if="$(arg use_odom_frame)" name="map_frame" value="map"/>
    <param unless="$(arg use_odom_frame)" name="map_frame" value="none"/>
    <param name="base_link_frame" value="base_link"/>
    <param name="world_frame" value="map"/>
 
</launch>
